controllers/release/release_adapter.go

```diff
@@ -229,25 +229,29 @@ func (a *Adapter) EnsureSnapshotEnvironmentBindingIsCreated() (results.Operation
 // EnsureTargetContextIsSet is an operation that will ensure that the targetContext of the adapter is not nil. It will
 // set it to a context from the workspace in the ReleasePlan if one is defined or to its own context if not.
 func (a *Adapter) EnsureTargetContextIsSet() (results.OperationResult, error) {
-	patch := client.MergeFrom(a.release.DeepCopy())
 	// If the target namespace/workspace is not set we should get the target from the ReleasePlan
 	if a.release.Status.Target == (kcp.NamespaceReference{}) {
+		patch := client.MergeFrom(a.release.DeepCopy())
 		releasePlan, err := a.getReleasePlan()
 		if err != nil {
 			a.release.MarkInvalid(v1alpha1.ReleaseReasonReleasePlanValidationError, err.Error())
 			return results.RequeueOnErrorOrStop(a.client.Status().Patch(a.context, a.release, patch))
 		}
+		if _, ok := logicalcluster.ClusterFromContext(a.context); !ok {
+			a.release.MarkInvalid(v1alpha1.ReleaseReasonReleasePlanValidationError, fmt.Sprintf("Workspace %s does not exist", releasePlan.Spec.Target.Workspace))
+			return results.RequeueOnErrorOrStop(a.client.Status().Patch(a.context, a.release, patch))
+		}
+		a.logger.Info("Running in a KCP environment")
 		a.release.Status.Target = releasePlan.Spec.Target
+		return results.RequeueOnErrorOrContinue(a.client.Status().Patch(a.context, a.release, patch))
 	}
 
 	if a.release.Status.Target.Workspace != "" {
```

Comment:
This is because it only enters that section if the `release.Status.Target` was not modified. So In case it was already set (from the releasePlan) inside that block it won't need to fetch the releasePlan once again, but from the release status.

Reply:
I don't follow. This is how I read the code
```
if the release.status.target is empty, which is only true the first time we reconcile a release
     get releaseplan
     if the releaseplan workspace is not empty (we are expecting kcp), see if the cluster exists
           if the cluster exists, output we are running in kcp to the logger. but don't set the status to that for some reason?
     set the release.status.target to the target of the releaseplan. so, the first if will never be true again for this release
now let's again see if a workspace was defined
     if it is, set the target context to the kcp one
     else, set it equal to the default context
```
I do not see why we cannot set the target context when we set the status

Reply:
I refactored this once again. Please check if now it makes more sense.

Reply:
Nope, let's meet about it